lib/ai_moderation/spam_scanner.rb

```diff
@@ -0,0 +1,349 @@
+        true
+      end
+
+      def self.scanned_max_times?(post)
+        AiSpamLog.where(post_id: post.id).count >= 3
+      end
+
+      def self.significant_change?(previous_version, current_version)
+        return true if previous_version.nil? # First edit should be scanned
+
+        # Use Discourse's built-in levenshtein implementation
+        distance =
+          ScreenedEmail.levenshtein(previous_version.to_s[0...1000], current_version.to_s[0...1000])
+
+        distance >= MINIMUM_EDIT_DIFFERENCE
+      end
+
+      def self.test_post(post, custom_instructions: nil)
+        settings = AiModerationSetting.spam
+        llm = settings.llm_model.to_llm
+        custom_instructions = custom_instructions || settings.custom_instructions.presence
+        context = build_context(post)
+        prompt = completion_prompt(post, context: context, custom_instructions: custom_instructions)
+
+        result =
+          llm.generate(
+            prompt,
+            temperature: 0.1,
+            max_tokens: 100,
+            user: Discourse.system_user,
+            feature_name: "spam_detection_test",
+            feature_context: {
+              post_id: post.id,
+            },
+          )&.strip
+
+        history = nil
+        AiSpamLog
+          .where(post: post)
+          .order(:created_at)
+          .limit(100)
+          .each do |log|
+            history ||= +"Scan History:\n"
+            history << "date: #{log.created_at} is_spam: #{log.is_spam}\n"
+          end
+
+        log = +"Scanning #{post.url}\n\n"
+
+        if history
+          log << history
+          log << "\n"
+        end
+
+        log << "LLM: #{settings.llm_model.name}\n\n"
+        log << "System Prompt: #{build_system_prompt(custom_instructions)}\n\n"
+        log << "Context: #{context}\n\n"
+        log << "Result: #{result}"
+
+        is_spam = check_if_spam(result)
+        { is_spam: is_spam, log: log }
+      end
+
+      def self.completion_prompt(post, context:, custom_instructions:)
+        system_prompt = build_system_prompt(custom_instructions)
+        prompt = DiscourseAi::Completions::Prompt.new(system_prompt)
+        args = { type: :user, content: context }
+        upload_ids = post.upload_ids
+        args[:upload_ids] = upload_ids.take(3) if upload_ids.present?
+        prompt.push(**args)
+        prompt
+      end
+
+      def self.perform_scan(post)
+        return if !enabled?
+        return if !should_scan_post?(post)
+
+        settings = AiModerationSetting.spam
+        return if !settings || !settings.llm_model
+
+        context = build_context(post)
+        llm = settings.llm_model.to_llm
+        custom_instructions = settings.custom_instructions.presence
+        prompt = completion_prompt(post, context: context, custom_instructions: custom_instructions)
+
+        begin
+          result =
+            llm.generate(
+              prompt,
+              temperature: 0.1,
+              max_tokens: 100,
```

Comment:
If we are not showing the LLM output with reasoning, we should lower this to 5 for performance reasons.